knowledge/knowledge_read.go

```diff
@@ -23,238 +23,226 @@ import (
-// that have the specified property.
-func (c *Client) ListResources(
-	ctx context.Context,
-	opts ...grpc.CallOption) ([]*knowledgepb.Node, error) {
-	return c.ListNodes(ctx, Resource, opts...)
-}
-
-// ListNodes is a helper function that lists all nodes by type, regardless of whether they are DigitalTwins
-// or Resources. The nodeType argument should be in PascalCase.
+// ListNodes is a helper function that lists all nodes by given node type, regardless of whether they are Identities
+// or not.
 func (c *Client) ListNodes(
 	ctx context.Context,
 	nodeType string,
 	opts ...grpc.CallOption,
-) ([]*knowledgepb.Node, error) {
+) ([]*knowledgeobjects.Node, error) {
 	ctx = insertMetadata(ctx, c.xMetadata)
-	path := fmt.Sprintf("(:%s)", nodeType)
-	resp, err := c.client.IdentityKnowledge(ctx, &knowledgepb.IdentityKnowledgeRequest{
-		Path:      path,
-		Operation: knowledgepb.Operation_OPERATION_READ,
+	query := fmt.Sprintf("MATCH (n:%s)", nodeType)
+	params := map[string]*objects.Value{}
+	resp, err := c.client.IdentityKnowledgeRead(ctx, &knowledgepb.IdentityKnowledgeReadRequest{
+		Query:       query,
+		InputParams: params,
+		Returns: []*knowledgepb.Return{
+			{
+				Variable: "n",
+			},
+		},
 	}, opts...)
 	if err != nil {
 		return nil, err
 	}
-	return parseMultipleNodesFromPaths(resp.GetPaths())
+	return resp.GetNodes(), nil
 }
 
 // ListNodesByProperty is a helper function that lists all nodes that have the specified type and property.
 func (c *Client) ListNodesByProperty(
 	ctx context.Context,
-	nodeType string,
-	property *knowledgepb.Property,
+	property *knowledgeobjects.Property,
+	isIdentity bool,
 	opts ...grpc.CallOption,
-) ([]*knowledgepb.Node, error) {
+) ([]*knowledgeobjects.Node, error) {
 	ctx = insertMetadata(ctx, c.xMetadata)
-	path := fmt.Sprintf("(n:%s)", nodeType)
-	conditions := fmt.Sprintf("WHERE n.%s = $%s", property.Key, property.Key)
-	params := make(map[string]*knowledgepb.InputParam)
-	switch property.Value.Value.(type) {
+	query := "MATCH (n:Resource)-[:HAS]->(p:Property)"
+	if isIdentity {
+		query = "MATCH (n:DigitalTwin)-[:HAS]->(p:Property)"
+	}
+	query = fmt.Sprintf(
+		"%s WHERE p.type='%s' and p.value=$%s",
+		query,
+		property.Type,
+		property.Type,
+	)
+
+	params := make(map[string]*objects.Value)
+	switch property.Value.Type.(type) {
 	case *objects.Value_IntegerValue:
-		params[property.Key] = &knowledgepb.InputParam{
-			Value: &knowledgepb.InputParam_IntegerValue{
+		params[property.Type] = &objects.Value{
+			Type: &objects.Value_IntegerValue{
 				IntegerValue: property.GetValue().GetIntegerValue()}}
 	case *objects.Value_StringValue:
-		params[property.Key] = &knowledgepb.InputParam{
-			Value: &knowledgepb.InputParam_StringValue{
+		params[property.Type] = &objects.Value{
+			Type: &objects.Value_StringValue{
 				StringValue: property.GetValue().GetStringValue()}}
 	default:
 		return nil, errors.New(codes.InvalidArgument, "only string or integer properties can be used for queries")
```

Comment:
we support more types than just `string` or `integer` now:
```
                        *objectspb.Value_StringValue,
			*objectspb.Value_BoolValue,
			*objectspb.Value_IntegerValue,
			*objectspb.Value_DoubleValue,
			*objectspb.Value_TimeValue,
			*objectspb.Value_DurationValue,
			*objectspb.Value_ArrayValue:
```

Reply:
added